### PR TITLE
New module: Add module for Amazon Egress Only Internet Gateway (cloud/amazon/ec2_vpc_egress_igw)

### PR DESCRIPTION
**Pull request by shepdelacreme**
##### SUMMARY

ec2_vpc_eigw: Module that creates or deletes an Egress Only Internet Gateway for the specified VPC. EIGW's are required to enable IPv6 instances in private subnets to communicate outbound to the internet. Operation of the EIGW API is very basic. Only a single gateway can be created for a VPC at any given time and there are no operations to attach or detach the gateway. Once created it is attached until it is deleted.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py

##### ANSIBLE VERSION

```
ansible 2.5.0 (add_ec2_vpc_egress_igw_module 3e5db71c43) last updated 2017/09/12 12:22:43 (GMT -400)
  config file = /Users/user/.ansible.cfg
  configured module search path = [u'/Users/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/user/devel/shep-ansible/lib/ansible
  executable location = /Users/user/devel/shep-ansible/bin/ansible
  python version = 2.7.13 (default, Apr  4 2017, 08:47:57) [GCC 4.2.1 Compatible Apple LLVM 8.1.0 (clang-802.0.38)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

**Comment by willthames** [comment]
I'd prefer this were better named `ec2_vpc_egress_gateway` or `ec2_vpc_egress_igw`

File: lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: You should probably catch this and then ignore it, and then handle it later through HAS_BOTO3

File: lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: I don't think this needs line breaking so heavily

File: lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: This line is over indented

File: lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: `if not response.get('EgressOnlyInternetGateways'):`

File: lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: I would consider adding a `wait` and `wait_timeout` attribute to this module and only wait if `wait` is true.

**Review by willthames — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py
Comment (shepdelacreme) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: Is there an example of this in another module? I did the above based on what I had seen in other modules. Also I'm importing HAS_BOTO3 from ansible.module_utils.ec2 already so would I stop doing that and just cut/paste the HAS_BOTO3 code from ec2.py into my module and add the botocore exceptions import?

Comment (shepdelacreme) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_eigw.py]: @willthames Disregard. I should have searched for my answer first before asking!

**Comment by shepdelacreme** [comment]
ready_for_review

**Review by willthames — APPROVED** [review]

**Comment by willthames** [comment]
shipit

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: `pause *= backoff`

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: Wouldn't this end up returning `changed=200` to callers? Usually we keep this a boolean.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: Seems this is only returned when state=present.

**Review by ryansb — CHANGES_REQUESTED** [review]
Minor tweaks, LGTM otherwise.

**Comment by calfonso** [comment]
@shepdelacreme waiting for your response to the review.

**Comment by bcoca** [comment]
removed shipit as this is still waiting for ryansb's review to be addressed

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (shepdelacreme) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: I had to review the boto3 docs to be sure but the delete_egress_only_internet_gateway function response syntax is as follows:
```
{
    'ReturnCode': True|False
}
```

http://boto3.readthedocs.io/en/latest/reference/services/ec2.html#EC2.Client.delete_egress_only_internet_gateway

So its returning a bool as ReturnCode. Seems counter-intuitive of the AWS API.

**Comment by shepdelacreme** [comment]
ready_for_review

**Comment by shepdelacreme** [comment]
Since this is a new module that won't be in until 2.5 now, is there a preference from the team on updating the module to use the new AWS Module utils outlined in
 https://github.com/ansible/community/blob/master/group-aws/utility-modules.md

i.e. Moving it over to the AnsibleAWSModule class?

**Comment by willthames** [comment]
@shepdelacreme for my latest new module, I made the conscious choice not to use AnsibleAWSModule, as this allows me to drop it into an Ansible 2.3 library. Once 2.4 is released (which I know isn't too far away), I'll revisit that. So it's up to you at the moment. We can always update it after merge.

**Comment by maraisr** [comment]
Be kinda nice to have this merged?

**Comment by shepdelacreme** [comment]
Made some updates to this module to use the new AnsibleAWSModule class and added integration tests. The integration tests should fail on shippable since I believe the IAM policy needs to have the following permissions added:

`ec2:CreateEgressOnlyInternetGateway`
`ec2:DeleteEgressOnlyInternetGateway`
`ec2:DescribeEgressOnlyInternetGateways`

**Comment by shepdelacreme** [comment]
ready_for_review

File: test/integration/targets/ec2_vpc_egress_igw/meta/main.yml
Comment (willthames) [inline test/integration/targets/ec2_vpc_egress_igw/meta/main.yml]: I don't think dependencies are needed

File: test/integration/targets/ec2_vpc_egress_igw/defaults/main.yml
Comment (willthames) [inline test/integration/targets/ec2_vpc_egress_igw/defaults/main.yml]: delete this file if it's effectively empty

File: test/integration/targets/ec2_vpc_egress_igw/tasks/main.yml
Comment (willthames) [inline test/integration/targets/ec2_vpc_egress_igw/tasks/main.yml]: We're using YAML anchors now to greatly simplify the AWS credential passing - see e.g. ec2_group tests.

File: test/integration/targets/ec2_vpc_egress_igw/tasks/main.yml
Comment (willthames) [inline test/integration/targets/ec2_vpc_egress_igw/tasks/main.yml]: Should always tear down the EIGW too (and ignore_errors in case it's not already created)

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: This check is no longer required with AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: Actually handled by AnsibleAWSModule

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: No need to import HAS_BOTO3 now.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: sample should just be the content of `gateway_id`, not the whole dict.

**Review by willthames — CHANGES_REQUESTED** [review]

**Review by willthames — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: BotoCoreError does not have a .response attribute so this may throw an AttributeError.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: The copyright can be updated: http://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#copyright

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (shepdelacreme) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: Ok I updated both sections that would be affected by this.

File: lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: e.message unfortunately doesn't exist at all on Python3. What you had before looked right. You could either break up 
```
try:
    boto_call
except ClientError as e:
    < things with e.response and fail_json_aws >
except BotoCoreError as e:
    < just fail_json_aws >
```
Or you could wrap the e.response in a `try`/`except AttributError`

Comment (s-hertel) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: I'm working on a PR to update CI for these tests right now. :)

Comment (shepdelacreme) [reply lib/ansible/modules/cloud/amazon/ec2_vpc_egress_igw.py]: ah ok I didn't test in Py3. I think I will split the excepts instead of adding another try/except. Thank you!

**Review by s-hertel — APPROVED** [review]

**Review by ryansb — APPROVED** [review]

**Pull request merged**
